applications/nrf_desktop/src/hw_interface/selector_hw.c

```diff
@@ -0,0 +1,308 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <zephyr.h>
+
+#include <device.h>
+#include <gpio.h>
+
+#include "selector_hw_def.h"
+
+#include "event_manager.h"
+#include "selector_event.h"
+#include "power_event.h"
+
+#define MODULE selector_hw
+#include "module_state_event.h"
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(MODULE, CONFIG_DESKTOP_SELECTOR_HW_LOG_LEVEL);
+
+static struct device       *gpio_dev[ARRAY_SIZE(port_map)];
+static struct gpio_callback gpio_cb[ARRAY_SIZE(port_map)];
+
+enum state {
+	STATE_DISABLED,
+	STATE_ACTIVE,
+	STATE_OFF,
+};
+
+static enum state state;
+static u8_t cur_selector_state = ARRAY_SIZE(pins);
+
+static struct k_delayed_work selector_work;
+
+
+static void selector_event_send(void)
+{
+	BUILD_ASSERT_MSG(ARRAY_SIZE(pins) <= UCHAR_MAX,
+			 "Maximum number of selector states is 255");
+	__ASSERT_NO_MSG(cur_selector_state < ARRAY_SIZE(pins));
+
+	struct selector_event *event = new_selector_event();
+
+	event->src = SELECTOR_SRC_HW;
+	event->state = cur_selector_state;
+
+	EVENT_SUBMIT(event);
+}
+
+static int read_state(void)
+{
+	u32_t value;
+
+	for (size_t i = 0; i < ARRAY_SIZE(pins); i++) {
+		int err = gpio_pin_configure(gpio_dev[pins[i].port],
+					     pins[i].pin,
+					     GPIO_PUD_PULL_DOWN | GPIO_DIR_IN);
+		if (err) {
+			LOG_ERR("Cannot configure pin (err %d)", err);
+			return err;
+		}
+
+		err = gpio_pin_read(gpio_dev[pins[i].port], pins[i].pin,
+				    &value);
+
+		if (err) {
+			LOG_ERR("Cannot read value from port:%" PRIu8
+				" pin: %" PRIu8, pins[i].port, pins[i].pin);
+			return err;
+		}
+
+		if (value) {
+			cur_selector_state = i;
+			return 0;
+		}
+	}
+
+	LOG_ERR("Switch state is unknown");
+	return -1;
+}
+
+static void selector_isr(struct device *dev, struct gpio_callback *cb,
+			 u32_t pins_mask)
+{
+	u8_t port = ARRAY_SIZE(port_map);
+
+	for (size_t i = 0; i < ARRAY_SIZE(port_map); i++) {
+		if (dev == gpio_dev[i]) {
+			port = i;
+			break;
+		}
+	}
+
+	__ASSERT_NO_MSG(port < ARRAY_SIZE(port_map));
+
+	for (size_t i = 0; i < ARRAY_SIZE(pins); i++) {
+		int err = gpio_pin_disable_callback(gpio_dev[pins[i].port],
+						    pins[i].pin);
+
+		if (err) {
+			LOG_ERR("Cannot disable callback (err %d)",
+				err);
+		}
+
+		if ((pins[i].port == port) && (pins_mask & BIT(pins[i].pin))) {
+			/* Only one pin may be set at the time. */
+			__ASSERT_NO_MSG((pins_mask & ~BIT(pins[i].pin)) == 0);
```

Comment:
If only one pin may be set at a time, is it enough to asser here? Isn't is exceptional situation that may (or at least should not happen)?

Reply:
For used switch this should not happen (maybe only when we have short circuit somewhere - faulty hardware). Assert is to make sure the code works correctly.

Reply:
So in case assertion fails it is a hard hw error and in my humble opinion is a serious problem...

Reply:
Ok, I will add setting module to error state here (as discussed offline).